### DISCUSSION TITLE
🔍 FAQ 问题汇总

### DISCUSSION THREAD

**Discussion by stdrc** [nonebot/discussions#13] [通用]
# 请不要在本帖提问，[提问请登录后点击此处](https://github.com/nonebot/discussions/discussions/new?category=%E9%97%AE%E7%AD%94)

> FAQ, 全称 **F**requently **A**sked **Q**uestions, 即"经常问到的问题"
顾名思义, 这里提供了一些在NoneBot开发过程中常见问题的解答

如果你觉得这里的回答有帮助, 请点击回答左下角给予支持
如果您有其他这里没有提及的常见问题想要补充, 也可以随时向这里进行补充

<details open="">
<summary><b>通用</b></summary>

- [Nonebot 文档使用的主题是什么？](#discussioncomment-738457)
- [Linux 怎么把 bot 挂到后台？](#discussioncomment-738458)
- [初始化时出现 ModuleNotFoundError: no module named xxx](#discussioncomment-738461)
- [ActionFailed（nonebot.adapters.cqhttp.exception.ActionFailed 或 aiocqhttp.exceptions.ActionFailed）](#discussioncomment-734514)
- [如何避免同步函数调用导致的事件循环阻塞?](#discussioncomment-1150734)
- [如何进行不会阻塞事件循环的网络请求?](#discussioncomment-1184822)

</details>

---

<details open="">
<summary><b>NoneBot v2</b></summary>

- [为什么更新后无法导入 adapter？](#discussioncomment-738465)
- [启动后无法连接上](#discussioncomment-738467)
- [启动后出现 TypeError/Type xxx 报错](#discussioncomment-738468)
- [定时任务如何获取 bot 实例](#discussioncomment-738470)
- [nb-cli 运行出错](#discussioncomment-738471)
- [为何 NoneBot 2 调用 `matcher.finish` 之类的方法发送的消息 CQ 码未解析/为纯文本？](#discussioncomment-744355)
- [如何在Windows下进行异步子进程调用(如使用Playwright等)?](#discussioncomment-1159147)
- [如何替换ASGI服务器](https://github.com/nonebot/discussions/discussions/13#discussioncomment-1394760)
- [NoneBot 2 的不同事件处理/逻辑控制函数之间究竟有何区别？](#discussioncomment-1447083)

</details>

---

<details open="">
<summary><b>NoneBot v1</b></summary>

</details>

**Comment by SK-415** [comment]
## ActionFailed（nonebot.adapters.onebot.exception.ActionFailed 或 aiocqhttp.exceptions.ActionFailed）

此问题与 NoneBot 无关，请参考协议端（如 go-cqhttp）的输出。

**Comment by bytemain** [comment]
QQ机器人如何发送图片？
QQ机器人如何发送文件？

**Reply by j1g5awi** [reply]
这应该是两个问题，因为发图片用的是 CQCode，发文件只有 go-cqhttp 支持而且是调用 API。

**Comment by j1g5awi** [comment]
## Nonebot 文档使用的主题是什么？

[`nonebot/docusaurus-theme-nonepress`](https://github.com/nonebot/docusaurus-theme-nonepress)

一款基于Docusaurus和React的文档主题, 由NoneBot团队自行研发 (当然, 你也可以用!)

**Comment by j1g5awi** [comment]
## Linux 怎么把 bot 挂到后台？

使用 shell `&` 运算符、[tmux](https://linuxize.com/post/getting-started-with-tmux/)、[screen](https://www.runoob.com/linux/linux-comm-screen.html)、[pm2](https://pm2.keymetrics.io/docs/usage/quick-start/)、[supervisord](http://supervisord.org/introduction.html) 等工具都可以将 bot 放到后台运行，请选择最适合自己的工具。

**Reply by jexjws** [reply]
有 `&` 运算符 的使用示例吗

**Reply by zty012** [reply]
> 有 `&` 运算符 的使用示例吗

我只知道powershell的

```bash
& nb run  # 挂后台
Get-Job  # 查看后台
Stop-Job <job-id>  # 杀后台
Remove-Job <job-id>  # 移除后台
```

**Reply by SuperKenVery** [reply]
也推荐一下用systemd（）

**Comment by j1g5awi** [comment]
## 初始化时出现 ModuleNotFoundError: no module named xxx

请先同时卸载 NoneBot 1 和 NoneBot 2：

```shell
pip uninstall nonebot
pip uninstall nonebot2
```

然后重新安装所需要的版本。

**为了保证依赖版本不会冲突，我们强烈推荐您使用[虚拟环境](https://www.liaoxuefeng.com/wiki/1016959663602400/1019273143120480)**

**Comment by j1g5awi** [comment]
## 为什么更新后无法导入 adapter？

![image](https://user-images.githubusercontent.com/13938334/118070564-be2f2280-b3d8-11eb-80ae-bae3bdaaa0c3.png)

报错：no module named cqhttp

解决方案：

```
pip install nonebot-adapter-cqhttp
```

在 2.0 的某一次更新中，把 adapter 拆了出来单独发包，导致了这个问题。

**Reply by anaivebird** [reply]
great

**Comment by j1g5awi** [comment]
## 启动后无法连接上

建议检查 host 与 port 配置，检查端口占用，检查 adapter 是否注册。

**Comment by j1g5awi** [comment]
## 启动后出现 TypeError/Type xxx 报错

请使用 Python 3.8+。

**Comment by j1g5awi** [comment]
## 定时任务如何获取 bot 实例

- `nonebot.get_bots()`

- 在`nonebot2.0.0a14`之后，我们添加了`nonebot.get_bot()`

**Comment by j1g5awi** [comment]
## nb-cli 运行出错

出现 `httpx` 字样时，多为无法获取 NoneBot 2 的云端资源（插件/bot/adapter 列表）或是代理服务器有问题，推荐切换或关闭代理后重试

请确保系统环境变量 `PATH` 中有 `pip` 的脚本目录

**Comment by j1g5awi** [comment]
## 升级 adapter 或 NoneBot 2 本体后运行出现报错

尝试同时升级 adapter 与本体。

**Comment by j1g5awi** [comment]
## 我的协议端是 go-cqhttp，该选择 cqhttp adapter 还是 gocq adapter？

gocq adapter 是非官方维护的适配器，对 go-cqhttp 的拓展 API 进行了适配，使得在 IDE 中可以获得更好的代码补全。
    
但如果你想获得**更稳定**的支持，请使用官方维护的 CQHTTP 适配器。

**Comment by StarHeartHunt** [comment]
## 为何 NoneBot 2 调用 `matcher.finish` 之类的方法发送的消息 CQ 码未解析/为纯文本？

为避免 SSRF 等注入问题，NoneBot 2 默认对使用这些方法发送的消息中的 CQ 码进行转义，作为纯文本发送，推荐使用 `MessageSegment` 构造 CQ 码消息段。

详情请见： https://github.com/nonebot/nonebot2/issues/212#issuecomment-778007483

**Comment by Evolvemomo** [comment]
提示找不到nb这个命令，已经安装了nb-cli
![image](https://user-images.githubusercontent.com/67366129/124092630-a8111a80-da89-11eb-89d6-0482da7feb4d.png)

**Reply by yanyongyu** [reply]
请去新开一个贴，不要在汇总贴询问

**Comment by mnixry** [comment]
## 如何避免同步函数调用导致的事件循环阻塞?

我们提供了以下工具函数, 通过[调用线程池](https://docs.python.org/zh-cn/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor)将任意同步步函数调用变为异步, 以避免主线程事件循环被阻塞:
  - 在NoneBot1中, 可以使用[`aiocqhttp.utils.ensure_async`](https://aiocqhttp.nonebot.dev/module/aiocqhttp/utils.html)
  - 在NoneBot2中, 可以使用[`nonebot.utils.run_sync`](https://v2.nonebot.dev/api/utils.html#run-sync-func)

以上函数的推荐用法是作为装饰器使用的

```python
@run_sync
def some_sync_function(a,b,c):
    do_something()

await some_sync_function(1,2,3)
```

如果你不想把它当作装饰器使用, 可以使用下面这种代码

```python
def some_sync_function(a,b,c):
    do_something()

await run_sync(some_sync_function)(1,2,3)
```

~~因为后者看起来很丑所以不推荐~~

source: https://github.com/nonebot/discussions/discussions/22

**Reply by cleoold** [reply]
不建議 `aiocqhttp.utils.ensure_async`。請自己使用綫程池或進程池再 `run_in_executor`，或使用 `quart.utils.run_sync`

**Reply by mnixry** [reply]
> 不建議 `aiocqhttp.utils.ensure_async`。請自己使用綫程池或進程池再 `run_in_executor`，或使用 `quart.utils.run_sync`

请解释原因并给出您所推荐的代码使用示例

**Comment by mnixry** [comment]
## 如何在Windows下进行异步子进程调用(如使用Playwright等)?

由于Uvicorn在Windows下选择的[事件循环(`SelectorEventLoop`)实现原因](https://docs.python.org/zh-cn/3/library/asyncio-subprocess.html), 它无法在Windows下异步地创建子进程 (相关Issue: https://github.com/nonebot/nonebot2/issues/376)

不过所幸, 目前我们可以通过[猴子补丁(`monkeypatch`)](https://zh.wikipedia.org/zh-hans/%E7%8C%B4%E8%A1%A5%E4%B8%81)来解决这个问题

- **注意: 启用这段代码后, 会在Windows平台下带来包括但不限于以下的副作用, 请酌情修改** 
  - `uvicorn`代码修改检测重载不再可用 
    - 相关链接: https://github.com/encode/uvicorn/pull/535
  - `aiohttp`的代理设置不再可用
    - 相关链接: https://github.com/aio-libs/aiohttp/issues/4536#issuecomment-579740877

> 以下代码参考自 @SK-415 https://github.com/nonebot/nonebot2/issues/376#issuecomment-846523618
> 从理论上, 如果在`nonebot.run()`执行之前导入, 这段代码将会起到强制Uvicorn使用**平台默认**的事件循环(在Windows下它是`ProactorEventLoop`)的作用
>
> ```python
> import asyncio
> import platform
> from uvicorn.loops import asyncio as _asyncio
> from uvicorn import config
> from nonebot.log import logger
> 
> 
> def asyncio_setup():
>     loop = asyncio.new_event_loop()
>     asyncio.set_event_loop(loop)
> 
> 
> @property
> def should_reload(self):
>     return False
> 
> 
> if platform.system() == "Windows":
>     _asyncio.asyncio_setup = asyncio_setup
>     config.Config.should_reload = should_reload
> ```

**Comment by FYWinds** [comment]
## 如何进行不会阻塞事件循环的网络请求？

因为NoneBot是一款异步框架，所以如果直接在代码中使用我们常用的`requests`或者`urllib3`等**同步**网络请求库，会导致事件循环阻塞，所以请在网络请求代码中使用异步的网络请求库：

**由于`aiohttp`在代理方面的实现在[跨平台方面兼容性较差](https://github.com/nonebot/discussions/discussions/13#discussioncomment-1159147)，推荐使用`httpx`**

- 如果举例来说，你可以将

  - 以下的同步请求
    ```python
    import requests
    
    def request_example():
        response = requests.get(url="https://www.example.com/)
        return response.content
    ```
  - 变成以下的异步请求
    ```python
    import httpx
    
    async def request_example():
        async with httpx.AsyncClient() as client:
            r = await client.get(url="https://www.example.com/")
        return r.content
    ```

- 更多的实例以及说明请移步[**httpx异步用法文档**(英文)](https://www.python-httpx.org/async/)
  - ~~如果有哪位大佬把这个文档翻译成中文贴在这个FAQ下就更好了~~

**Comment by synodriver** [comment]
## 如何替换ASGI服务器

uvicorn~~太拉跨了~~不支持某些高级特性，比如http2之类的，这样可以换个先进的ASGI服务器

- 使用[hypercorn](https://github.com/pgjones/hypercorn)，功能最丰富的ASGI服务器，**在需要http2支持以及asgi extension支持的时候使用**

  ```python
  # file: bot.py
  import asyncio
  
  from pathlib import Path
  
  import nonebot
  from nonebot.adapters.cqhttp import Bot
  
  nonebot.init(_env_file=r".env")
  driver = nonebot.get_driver()
  driver.register_adapter("cqhttp", Bot)
  
  nonebot.load_plugins(str(Path(".") / "plugins"))
  app = nonebot.get_app()
  
  if __name__ == "__main__":
      from hypercorn.asyncio import serve
      from hypercorn.config import Config
      try:
          import uvloop
          uvloop.install()
      except ImportError:
          pass
  
      config = Config()
      config.bind = ["0.0.0.0:8090"]
      config.accesslog = "-"
      config.loglevel = "DEBUG"
      # config.keyfile ="./key.pem"
      # config.certfile="./cert.pem"
      asyncio.run(serve(app, config))
  
  ```

- 使用[NGINX UNIT](https://github.com/nginx/unit)，最快的ASGI服务器，在需要**极致性能**的时候使用（比如我同时1000个群，数十个bot）
  - [官方文档提供的配置参考](http://unit.nginx.org/configuration/)

  - 配置文件`config.json`:
  ```json
  {
    "listeners": {
      "*:9000": {
        "pass": "applications/nonebot"
      }
    },
    "applications": {
      "nonebot": {
        "type": "python 3.8",
        "path": "你的bot.py的路径",
        "module": "bot",
        "callable": "app",
          "processes": 1
      }
    }
  }
  ```
  - 然后执行以下指令更新UNIT配置
  ```shell
  curl -X PUT --data-binary @config.json --unix-socket /path/to/control.unit.sock http://localhost/config/
  ```

**Comment by jks15satoshi** [comment]
## NoneBot 2 的不同事件处理/逻辑控制函数之间究竟有何区别？

### 阅读前须知：

本 FAQ 部分内容仅适用于 **v2.0.0a16 及之前** 的版本。部分 v2.0.0b1 新特性内容暂未更新 ~~（特性熟悉清楚后再更）~~，未在本 FAQ 中体现，请以官方文档为准。敬请留意。

---

NoneBot 2 提供了多个事件处理函数与逻辑控制函数以应对多种实际需求，尽管在文档中已经有章节解释了这些函数的用法以及各自的区别，但对于许多新人来说仍然是一个不易理解的概念。因此这里对文档进行一下汇总和精炼，尽可能通俗地解释这些函数的作用与异同。

*注：这则 FAQ 中不会涉及到 **完整的** 事件处理流程，而仅从开发者视角解释事件处理，所以会忽略一些实际流程中的部分细节。如果想了解详细的流程，请移步至 [该文档](https://v2.nonebot.dev/advanced/#%E5%AE%83%E5%A6%82%E4%BD%95%E5%B7%A5%E4%BD%9C)。* ~~*另外该 FAQ 面向新人开发者，本人也是菜鸟一个，大佬请轻喷。*~~

### 前置：响应器与事件处理函数

**响应器** (matcher) 定义 bot 响应符合条件的事件，通俗来讲，你在插件中使用的 `on_*` 函数就是在注册响应器[^1]。Bot 会在接收到你所定义的事件类型且满足 rule 与 permission 条件后触发响应器，执行响应器中定义的逻辑。

**事件处理函数** (或处理器，handler) 定义 bot 响应事件后执行的具体逻辑。它与响应器是包含关系，一个响应器中可以定义多个事件处理函数。当响应器被触发时，会 **依次** 执行响应器中定义的事件处理函数[^2]。

### 事件处理函数装饰器

NoneBot 2 提供了三种不同的事件处理函数装饰器，不同装饰器对应不同的事件处理方式。使用时都需要将其装饰在一个 [符合规定定义](https://v2.nonebot.dev/guide/creating-a-handler.html#%E4%BA%8B%E4%BB%B6%E5%A4%84%E7%90%86%E5%87%BD%E6%95%B0%E5%8F%82%E6%95%B0) 的逻辑函数上，而被装饰器装饰的逻辑函数则会被视为事件处理函数被添加到响应器中。

> 为了便于后面的阐述，关于事件处理函数，这里插一嘴简单介绍一下。
>
> 通常来讲，事件处理函数包括以下三个参数（具体定义请阅读上面链接的文档）：`bot`、`event` 以及 `state`：
>
> - `bot` 是响应当前事件的 bot 实例，在执行处理器过程中你可以通过这个参数执行例如 API 调用等操作；
> - `event` 是当前事件的事件对象，里面包括了当前事件的一些数据，例如消息发送者等[^3]；
> - `state` 是当前事件的状态字典，里面预存了一些由 NoneBot 处理的信息，也可以按需存放事件处理过程中需要的信息，以便实现诸如用户交互等场景。

这三种装饰器的区别在于：

- `handle()`：最基本的事件处理函数装饰器，其作用仅仅是向响应器添加一个事件处理器。用于纯粹的事件处理逻辑、没有复杂的逻辑交互的场景；也可与其他装饰器配合，用于处理一个固定的流程；
- `receive()`：被该装饰器装饰的处理函数将等待用户发送一条新消息后执行处理逻辑；如果该处理函数是响应器中第一个处理函数，则直接使用第一条接收到的消息。*（由于本人没有使用过该装饰器因此应用场景有待补充）*；
- `got(key, prompt, args_parser)`：被该装饰器装饰的处理函数将在事件状态字典  `state` 中检查是否存在 `key`，如果存在则执行该处理函数，否则等待用户发送一条新消息后执行，同时发送由 `prompt` 定义的消息指示用户发送一条新消息。`args_parser` 定义当该处理函数接收到新消息后如何将消息处理为所需的信息存入状态字典。常见的应用场景即为需要用户交互补充额外信息的情况。

你可能在文档中看到事件处理的示例代码中是先定义了 `handle` 事件处理函数，然后定义了 `got` 事件处理函数[^4]。但事实上这并不是硬性的流程，你完全可以一开始就定义 `got`[^5] 或是其他的事件处理函数，也可以定义任意数量的事件处理函数……总之，一切都取决于你的实际处理流程的需要。

### 事件处理逻辑控制函数

除了装饰器以外，另一个可能让新人开发者感到疑惑的就是逻辑控制函数。NoneBot 2 为响应器提供了四种逻辑控制函数，以便开发者灵活控制事件处理的流程。

> 开始前请留意之前所提到的，响应器中多个事件处理函数是 **顺序执行** 的。

所有的逻辑控制函数都有发送消息的行为（其中 `**kwargs` 参数会直接传递给 [`bot.send`](https://v2.nonebot.dev/api/adapters/#abstract-async-send-event-message-kwargs) 函数），区别在于：

- `matcher.send`：直接发送一条 `message` 消息[^6]；
- `matcher.pause`：结束当前的事件处理函数，等待用户发送一条消息后 **执行下一个事件处理函数**；
- `matcher.reject`：结束当前的事件处理函数，等待用户发送一条消息后 **重新执行当前的事件处理函数**；
- `matcher.finish`：结束整个事件处理，等待下一次事件响应。

通过上面的逻辑控制函数，你就可以更灵活地设计用户交互与事件处理流程。

### 小结

以上就是对事件处理函数装饰器与逻辑控制函数的粗略解释 ~~（复读文档）~~。更加详细的内容依然建议阅读 NoneBot 2 的相关文档，尤其是 [进阶文档](https://v2.nonebot.dev/advanced/#%E5%AE%83%E5%A6%82%E4%BD%95%E5%B7%A5%E4%BD%9C)，里面详细解释了 NoneBot 2 的完整工作流程，对 bot 功能的开发会更有帮助。

---

### Changelog

2022/01/16: 备注过时须知。

[^1]: 你可以在 [这个 API 文档](https://v2.nonebot.dev/api/plugin.html#on-type-rule-none-permission-none-handlers-none-temp-false-priority-1-block-false-state-none-state-factory-none) 中找到所有的事件响应器类型。
[^2]: 参考 NoneBot 2 文档 [处理 event](https://v2.nonebot.dev/advanced/#%E5%A4%84%E7%90%86-event) 章节。
[^3]: 对于 CQHTTP 适配器，你可以在 [这篇文档](https://github.com/botuniverse/onebot/blob/master/v11/specs/event/README.md) 中查看所有 OneBot 协议支持的事件类型与事件定义；同时如果你所使用的协议端定义的事件与标准 OneBot 协议有出入，你可能还需要参考你所使用的协议端的文档，例如 [GO-CQHTTP 文档 - 事件](https://docs.go-cqhttp.org/event/)。实际使用时以协议端定义为准。
对于其他适配器，可以参考对应平台的开发者文档。
[^4]: 参考 NoneBot 2 文档 [事件处理函数](https://v2.nonebot.dev/guide/creating-a-handler.html#%E4%BA%8B%E4%BB%B6%E5%A4%84%E7%90%86%E5%87%BD%E6%95%B0) 章节。
[^5]: 可以参考 [该示例](https://github.com/jks15satoshi/echobot/blob/main/echobot/plugins/botman/reboot.py)。~~BTW，这里打一下我个人项目的广告。~~
[^6]: 正如 [文档](https://v2.nonebot.dev/guide/creating-a-handler.html#matcher-send) 所说，这个函数严格来讲并不能算逻辑控制函数。不同于其他三个函数，这个函数除了发送一条消息以外，并没有任何事件处理流程上的控制，只能算是一种快捷操作。

**Comment by kentelove** [comment]
请问下,从群聊中踢人,要调用哪个API?

**Reply by yanyongyu** [reply]
请不要在汇总帖进行提问

如果是onebot v11的话API是 `set_group_kick`

**Comment by synodriver** [comment]
## 为什么onebot的客户端莫名其妙的unexpected EOF ~~记录一下刚才遇到的坑~~

属于ASGI服务器配置问题。千万不要设置读取超时时间，可能的话，让ASGI服务器自己定时ping。如果设置了读取超时，则一旦客户端长时间不上报数据就会被服务器主动关闭连接。

**Comment by zyp981204** [comment]
device lock is disable.设备锁未打开，请问是配置表少改了什么么？

**Reply by yanyongyu** [reply]
请不要在汇总帖进行提问

设备锁在QQ设置-帐号安全-登录设备管理

**Comment by zyp981204** [comment]
好的，没注意是汇总贴，抱歉




风泊
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "nonebot/discussions"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年6月16日(星期四) 上午10:13
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [nonebot/discussions] 🔍 FAQ 问题汇总 (Discussion #13)





 
请不要在汇总帖进行提问
 
设备锁在QQ设置-帐号安全-登录设备管理

**Comment by abandeng** [comment]
我安装成功了但是后面创建的时候提示nb不是内部命令

**Reply by yanyongyu** [reply]
请不要在汇总帖进行提问

脚手架使用方法或问题请参考

- https://v2.nonebot.dev/docs/next/start/nb-cli
- https://github.com/nonebot/discussions/discussions/13#discussioncomment-738471

**Comment by zty012** [comment]
## 控制台里全是心跳包事件？

关闭心跳包或者修改log-level为debug以上
